cluster-autoscaler/cloudprovider/mcm/mcm_manager.go

```diff
@@ -602,43 +603,109 @@ func (m *McmManager) retry(fn func(ctx context.Context) (bool, error), resourceT
-func (m *McmManager) GetMachineDeploymentNodes(machinedeployment *MachineDeployment) ([]string, error) {
-	md, err := m.machineDeploymentLister.MachineDeployments(m.namespace).Get(machinedeployment.Name)
-	if err != nil {
-		return nil, fmt.Errorf("Unable to fetch MachineDeployment object %s, Error: %v", machinedeployment.Name, err)
-	}
+// GetInstancesForMachineDeployment returns list of cloudprovider.Instance for machines which belongs to the MachineDeployment.
+func (m *McmManager) GetInstancesForMachineDeployment(machinedeployment *MachineDeployment) ([]cloudprovider.Instance, error) {
+	var (
+		list     = []string{machinedeployment.Name}
+		selector = labels.NewSelector()
+		req, _   = labels.NewRequirement("name", selection.Equals, list)
+	)
 
-	machineList, err := m.machineLister.Machines(m.namespace).List(labels.Everything())
+	selector = selector.Add(*req)
+	machineList, err := m.machineLister.Machines(m.namespace).List(selector)
 	if err != nil {
-		return nil, fmt.Errorf("Unable to fetch list of Machine objects %v", err)
+		return nil, fmt.Errorf("unable to fetch list of Machine objects %v for machinedeployment %q", err, machinedeployment.Name)
 	}
 
 	nodeList, err := m.nodeLister.List(labels.Everything())
 	if err != nil {
-		return nil, fmt.Errorf("Unable to fetch list of Nodes %v", err)
+		return nil, fmt.Errorf("unable to fetch list of Nodes %v", err)
 	}
 
-	var nodes []string
+	var instances []cloudprovider.Instance
 	// Bearing O(n2) complexity, assuming we will not have lot of nodes/machines, open for optimisations.
 	for _, machine := range machineList {
-		if strings.Contains(machine.Name, md.Name) {
-			var found bool
-			for _, node := range nodeList {
-				if machine.Labels["node"] == node.Name {
-					nodes = append(nodes, node.Spec.ProviderID)
-					found = true
-					break
+		instance := cloudprovider.Instance{}
+		var found bool
+		for _, node := range nodeList {
+			if machine.Labels["node"] == node.Name {
+				instance.Id = node.Spec.ProviderID
+				found = true
+				break
+			}
+		}
+		if !found {
+			// No k8s node found - either the VM is up but has not registered yet or registered but node is NotReady or MCM is unable to fulfill the request to create VM.
+			// Report a special placeholder ID so that the autoscaler can track it as an unregistered node.
+			instance.Id = placeholderInstanceIDForMachineObj(machine.Name)
+			instance.Status = &cloudprovider.InstanceStatus{
+				State:     cloudprovider.InstanceCreating,
+				ErrorInfo: getCreateErrorInfo(machine),
+			}
+		}
+		instances = append(instances, instance)
+	}
+	return instances, nil
+}
+
+func placeholderInstanceIDForMachineObj(name string) string {
+	return fmt.Sprintf("requested://%s", name)
+}
+
+// getCreateErrorInfo returns cloudprovider.InstanceErrorInfo for the machine obj
+func getCreateErrorInfo(machine *v1alpha1.Machine) *cloudprovider.InstanceErrorInfo {
```

Comment:
The name of this function suggests that it will create an ErrorInfo for any error that it sees in the status of the `Machine` object passed as an argument, however it only checks for `ResourceExhausted` errors. For all other errors it will return nil. The method name should indicate that.

Alternative name: `checkAndGetResourceExhaustedError`

Reply:
It might be extended in the future to also set `ErrorInfo` for other cases so the name seems OK.

Reply:
When you extend it then you change the name, till then please name this correctly.